lib/pbench/test/functional/agent/cli/commands/tools/test_clear_tools.py

```diff
@@ -70,6 +77,8 @@ def test_clear_tools_test65(monkeypatch, agent_config, pbench_run, pbench_cfg):
     assert pidstat_tool.exists() is False
     assert turbostat_tool.exists() is False
     assert mpstat_tool.exists() is False
+    # FIXME:  Why do we want to retain the group directory after deleting everything out of it??
+    # FIXME:  Contrary to the test docstring, this test doesn't create or assert that the default group is unmolested.
```

Comment:
These be good questions, matie. Can we fix them? 👍🏻

Reply:
Before I can fix the first one, I need to know whether it is intended behavior...and perhaps _why_....

As for fixing the second one...well, originally, my intention was to make as few changes to the unit tests as possible, to "show" that my changes to the CUT were "good" (i.e., if the new code works with the old test, then it must be OK, right??), so I didn't want to step too far out on that slippery slope...but, y'know, I'm sure that changing comments like docstrings would be _harmless_...of course.  😆

Reply:
About the first one I don't think it is an intended behavior I think it's a bug (I dont see how it can be an intended behavior since we can always create a same directory again if it doesn't exist). We should always create tools directory on demand and shouldn't leave any garbage on disk when a user clears those tools.

Reply:
What I find curious is how awkwardly the tools behave when the `default` group doesn't exist.

In point of fact, the `default` group is no different from any other group -- it's just the group which gets used when no other group is specified -- but there is a difference in behavior between dealing with a group which has never existed (this is a "Bad tool group") and a group which once existed but has since been cleared (which is silently ignored).  So, if you've never had a `default` group, the first time you run these commands, you'll get something like this:
```
# pbench-list-tools
No tool groups found
# pbench-clear-tools 
pbench-clear-tools: invalid --group option "default" (Bad tool group, 'default': directory /var/lib/pbench-agent/tools-v1-default does not exist)
# pbench-register-tool-set 
"vmstat" tool is now registered for host "401c3456f30f" in group "default"
"iostat" tool is now registered for host "401c3456f30f" in group "default"
"sar" tool is now registered for host "401c3456f30f" in group "default"
# pbench-clear-tools 
Removed "iostat" from host "401c3456f30f" in tools group "default"
Removed "sar" from host "401c3456f30f" in tools group "default"
Removed "vmstat" from host "401c3456f30f" in tools group "default"
All tools removed from group "default" on host "401c3456f30f"
# :
# pbench-list-tools
No tools found
# pbench-clear-tools 
No such group "default".
```
Note the difference between the first invocation of `pbench-list-tools` and the last, and between the first invocation of `pbench-clear-tools` and the last.

I'm inclined to defer changing the behavior with respect to removing the tool directories to another PR (perhaps the one addressing #2337?  😁), but I agree that I think we should be removing them...but I think we should handle the situation when `default` is missing a little more gracefully, even if it requires some special-case code.  In the contest of that change, it might be appropriate to add some assertions to the unit tests.

Reply:
Yes It can be a part of #2337 since you pointed this bug under that issue. :)